lib/texd/cache.rb

```diff
@@ -0,0 +1,102 @@
+# frozen_string_literal: true
+
+module Texd
+  # Cache is a simple LRU cache with a double linked list.
+  # If the cache is full the last element from the list will be removed.
+  class Cache
+    class LinkedList
+      attr_reader :head
+
+      def add(node)
+        if @head
+          node.next       = @head
+          node.prev       = @head.prev
+          @head.prev.next = node
+          @head.prev      = node
+        else
+          node.next = node
+          node.prev = node
+        end
+
+        @head = node
+      end
+
+      def remove(node)
+        if node.next == node
+          # remove last element, should never happen
+          @head = nil
+        else
+          node.prev.next = node.next
+          node.next.prev = node.prev
+          @head          = node.next if node == @head
+        end
+      end
+    end
+
+    class Node
+      attr_reader :key
+      attr_accessor :value, :prev, :next
+
+      def initialize(key)
+        @key = key
+      end
+    end
+
+    attr_reader :list, :hash, :count, :capacity
+
+    def initialize(capacity)
+      @list     = LinkedList.new
+      @hash     = {}
+      @mtx      = Mutex.new
+      @count    = 0
+      @capacity = capacity
+    end
+
+    def lookup(key)
+      read(key) || write(key, yield)
+    end
```

Comment:
I've added `#lookup`, it makes using the cache a bit more idiomatic:

```rb
cache.lookup(:foo) { "expensive computation" }
```

I'd like a better name though, as it performs a "lookup or store" action. The syntax is inspired by Rails' fragment caching, where we can do

```erb
<% cache "key" do %>
  <p>HTML fragment</p>
<% end %>
```

Re-using the name "cache" feels wrong on the call site though:

```rb
cache.cache(:foo) { "expensive computation" }
```

Any opinions?

Reply:
Good idea!

Reply:
We're going with `#fetch`.